File: halatuju-web/src/lib/pathways.ts
/**
 * Pathway eligibility engine for Matriculation and STPM (Form 6).
 *
 * Runs entirely on the frontend. No backend calls needed.
 * Input: student's SPM grades + CoQ score.
 * Output: eligibility + merit/score per pathway track.
 */

// --- Grade Point Scales ---

// Matriculation merit scale (from matrikulasi.moe.gov.my calculator)
const MATRIC_GRADE_POINTS: Record<string, number> = {
  'A+': 25, 'A': 24, 'A-': 23, 'B+': 22, 'B': 21,
  'C+': 20, 'C': 19, 'D': 18, 'E': 17, 'G': 0,
}

// STPM (Form 6) mata gred scale — lower is better
const STPM_MATA_GRED: Record<string, number> = {
  'A+': 1, 'A': 1, 'A-': 2, 'B+': 3, 'B': 4,
  'C+': 5, 'C': 6, 'D': 7, 'E': 8, 'G': 9,
}

// Credit = C or better (mata gred <= 6)
function isCredit(grade: string): boolean {
  const mg = STPM_MATA_GRED[grade]
  return mg !== undefined && mg <= 6
}

// Grade meets minimum threshold
function meetsMin(grade: string, minGrade: string): boolean {
  const order = ['A+', 'A', 'A-', 'B+', 'B', 'C+', 'C', 'D', 'E', 'G']
  const gradeIdx = order.indexOf(grade)
  const minIdx = order.indexOf(minGrade)
  if (gradeIdx === -1 || minIdx === -1) return false
  return gradeIdx <= minIdx // lower index = better grade
}

// --- Matriculation ---

export interface MatricTrack {
  id: string
  name: string
  nameMs: string
  nameTa: string
}

export const MATRIC_TRACKS: MatricTrack[] = [
  { id: 'sains', name: 'Science', nameMs: 'Sains', nameTa: 'அறிவியல்' },
  { id: 'kejuruteraan', name: 'Engineering', nameMs: 'Kejuruteraan', nameTa: 'பொறியியல்' },
  { id: 'sains_komputer', name: 'Computer Science', nameMs: 'Sains Komputer', nameTa: 'கணினி அறிவியல்' },
  { id: 'perakaunan', name: 'Accounting', nameMs: 'Perakaunan', nameTa: 'கணக்கியல்' },
]

interface MatricRequirement {
  subjectId: string
  minGrade: string
  alternatives?: string[] // alternative subject IDs (pick one)
}

// Subject requirements per track
// Subject IDs match the grades page: MAT, AMT, CHE, PHY, BIO, COMP_SCI, ACC, etc.
const MATRIC_REQUIREMENTS: Record<string, MatricRequirement[]> = {
  sains: [
    { subjectId: 'MAT', minGrade: 'B' },
    { subjectId: 'AMT', minGrade: 'C' },
    { subjectId: 'CHE', minGrade: 'C' },
    { subjectId: 'PHY', minGrade: 'C', alternatives: ['BIO'] },
  ],
  kejuruteraan: [
    { subjectId: 'MAT', minGrade: 'B' },
    { subjectId: 'AMT', minGrade: 'C' },
    { subjectId: 'PHY', minGrade: 'C' },
    // 4th: any elective with min C — handled separately
  ],
  sains_komputer: [
    { subjectId: 'MAT', minGrade: 'C' },
    { subjectId: 'AMT', minGrade: 'C' },
    { subjectId: 'COMP_SCI', minGrade: 'C' },
    // 4th: any elective with min C — handled separately
  ],
  perakaunan: [
    { subjectId: 'MAT', minGrade: 'C' },
    // 3 electives with min C — handled separately
  ],
}

export interface PathwayResult {
  pathway: 'matric' | 'stpm'
  trackId: string
  trackName: string
  trackNameMs: string
  trackNameTa: string
  eligible: boolean
  merit?: number         // Matric merit score (0-100)
  mataGred?: number      // STPM total mata gred
  maxMataGred?: number   // STPM threshold
  reason?: string        // Why not eligible (i18n key)
  reasonParams?: Record<string, string>
}

function findBestElective(
  grades: Record<string, string>,
  excludeIds: Set<string>,
  minGrade: string
): { id: string; grade: string } | null {
  let best: { id: string; grade: string; pts: number } | null = null
  for (const [id, grade] of Object.entries(grades)) {
    if (excludeIds.has(id)) continue
    if (!meetsMin(grade, minGrade)) continue
    const pts = MATRIC_GRADE_POINTS[grade] || 0
    if (!best || pts > best.pts) {
      best = { id, grade, pts }
    }
  }
  return best ? { id: best.id, grade: best.grade } : null
}

function checkMatricTrack(
  trackId: string,
  grades: Record<string, string>,
  coqScore: number
): PathwayResult {
  const track = MATRIC_TRACKS.find(t => t.id === trackId)!
  const reqs = MATRIC_REQUIREMENTS[trackId]
  const usedSubjects: { id: string; grade: string }[] = []
  const usedIds = new Set<string>()

  // Check fixed requirements
  for (const req of reqs) {
    const candidates = [req.subjectId, ...(req.alternatives || [])]
    let found = false
    for (const subjId of candidates) {
      const grade = grades[subjId]
      if (grade && meetsMin(grade, req.minGrade)) {
        usedSubjects.push({ id: subjId, grade })
        usedIds.add(subjId)
        found = true
        break
      }
    }
    if (!found) {
      // Check if student has the subject but grade too low
      const hasSubject = candidates.some(id => grades[id])
      const subjectName = candidates[0]
      return {
        pathway: 'matric',
        trackId,
        trackName: track.name,
        trackNameMs: track.nameMs,
        trackNameTa: track.nameTa,
        eligible: false,
        reason: hasSubject ? 'pathways.gradeTooLow' : 'pathways.subjectMissing',
        reasonParams: { subject: subjectName, minGrade: req.minGrade },
      }
    }
  }

  // Fill remaining slots with best electives (for tracks that need them)
  const slotsNeeded = 4 - usedSubjects.length
  for (let i = 0; i < slotsNeeded; i++) {
    const minGrade = 'C'
    const elective = findBestElective(grades, usedIds, minGrade)
    if (!elective) {
      return {
        pathway: 'matric',
        trackId,
        trackName: track.name,
        trackNameMs: track.nameMs,
        trackNameTa: track.nameTa,
        eligible: false,
        reason: 'pathways.notEnoughElectives',
      }
    }
    usedSubjects.push(elective)
    usedIds.add(elective.id)
  }

  // Calculate merit
  const subjectPoints = usedSubjects.reduce(
    (sum, s) => sum + (MATRIC_GRADE_POINTS[s.grade] || 0), 0
  )
  const academic = (subjectPoints / 100) * 90
  const coq = Math.min(Math.max(coqScore, 0), 10)
  const merit = Math.min(academic + coq, 100)

  return {
    pathway: 'matric',
    trackId,
    trackName: track.name,
    trackNameMs: track.nameMs,
    trackNameTa: track.nameTa,
    eligible: true,
    merit: Math.round(merit * 100) / 100,
  }
}

// --- STPM (Form 6) ---

// Subject groups for STPM Science bidang
// Student must have credits from 3 DIFFERENT groups
const STPM_SCIENCE_GROUPS: string[][] = [
  ['MAT', 'AMT'],
  ['PHY'],
  ['CHE'],
  ['BIO'],
  ['ENG_DRAW', 'ENG_MECH', 'ENG_CIVIL', 'ENG_ELEC', 'REKA_CIPTA',
   'SPORTS_SCI', 'SRT', 'COMP_SCI', 'GKT'],
]

// Subject groups for STPM Social Science bidang
const STPM_SOCSCI_GROUPS: string[][] = [
  ['BM'],
  ['BI'],
  ['SEJ'],
  ['GEO', 'PSV'],
  ['PI', 'PM'],
  ['MAT', 'AMT'],
  ['ACC'],
  ['SN', 'ADDSCI'],
  ['ECO', 'BUS', 'KEUSAHAWANAN', 'SPORTS_SCI', 'SRT', 'COMP_SCI',
   'GKT', 'PERTANIAN'],
]

export interface StpmBidang {
  id: string
  name: string
  nameMs: string
  nameTa: string
  maxMataGred: number
}

export const STPM_BIDANGS: StpmBidang[] = [
  { id: 'sains', name: 'Science', nameMs: 'Sains', nameTa: 'அறிவியல்', maxMataGred: 18 },
  { id: 'sains_sosial', name: 'Social Science', nameMs: 'Sains Sosial', nameTa: 'சமூக அறிவியல்', maxMataGred: 12 },
]

function checkStpmBidang(
  bidangId: string,
  grades: Record<string, string>
): PathwayResult {
  const bidang = STPM_BIDANGS.find(b => b.id === bidangId)!
  const groups = bidangId === 'sains' ? STPM_SCIENCE_GROUPS : STPM_SOCSCI_GROUPS

  // General requirement: credit in BM
  const bmGrade = grades['BM']
  if (!bmGrade || !isCredit(bmGrade)) {
    return {
      pathway: 'stpm',
      trackId: bidangId,
      trackName: bidang.name,
      trackNameMs: bidang.nameMs,
      trackNameTa: bidang.nameTa,
      eligible: false,
      maxMataGred: bidang.maxMataGred,
      reason: 'pathways.bmCreditRequired',
    }
  }

  // Find best 3 credits from different groups
  // Greedy: try each group, pick the subject with lowest mata gred
  type Pick = { groupIdx: number; subjectId: string; mataGred: number }
  const candidates: Pick[] = []

  for (let gi = 0; gi < groups.length; gi++) {
    let bestInGroup: Pick | null = null
    for (const subjId of groups[gi]) {
      const grade = grades[subjId]
      if (!grade || !isCredit(grade)) continue
      const mg = STPM_MATA_GRED[grade]
      if (!bestInGroup || mg < bestInGroup.mataGred) {
        bestInGroup = { groupIdx: gi, subjectId: subjId, mataGred: mg }
      }
    }
    if (bestInGroup) candidates.push(bestInGroup)
  }

  // Sort by mata gred (lowest first = best)
  candidates.sort((a, b) => a.mataGred - b.mataGred)

  if (candidates.length < 3) {
    return {
      pathway: 'stpm',
      trackId: bidangId,
      trackName: bidang.name,
      trackNameMs: bidang.nameMs,
      trackNameTa: bidang.nameTa,
      eligible: false,
      maxMataGred: bidang.maxMataGred,
      reason: 'pathways.notEnoughCredits',
      reasonParams: { needed: '3', have: String(candidates.length) },
    }
  }

  // Take best 3
  const best3 = candidates.slice(0, 3)
  const totalMataGred = best3.reduce((sum, p) => sum + p.mataGred, 0)

  if (totalMataGred > bidang.maxMataGred) {
    return {
      pathway: 'stpm',
      trackId: bidangId,
      trackName: bidang.name,
      trackNameMs: bidang.nameMs,
      trackNameTa: bidang.nameTa,
      eligible: false,
      mataGred: totalMataGred,
      maxMataGred: bidang.maxMataGred,
      reason: 'pathways.mataGredTooHigh',
      reasonParams: { total: String(totalMataGred), max: String(bidang.maxMataGred) },
    }
  }

  return {
    pathway: 'stpm',
    trackId: bidangId,
    trackName: bidang.name,
    trackNameMs: bidang.nameMs,
    trackNameTa: bidang.nameTa,
    eligible: true,
    mataGred: totalMataGred,
    maxMataGred: bidang.maxMataGred,
  }
}

// --- Prestige Scoring ---
// Pre-university pathways get a prestige bonus + academic bonus
// so they compete fairly with quiz-boosted courses (max 120).

const PRESTIGE_BONUS = 8
const BASE_SCORE = 100

function matricAcademicBonus(merit: number): number {
  if (merit >= 92) return 8
  if (merit >= 87) return 5
  if (merit >= 82) return 3
  return 0 // below 82 = not eligible (shouldn't reach here)
}

function stpmAcademicBonus(mataGred: number, bidangId: string): number {
  if (bidangId === 'sains') {
    if (mataGred <= 6) return 8
    if (mataGred <= 10) return 5
    if (mataGred <= 14) return 3
    if (mataGred <= 18) return 1
    return 0
  }
  // sains_sosial
  if (mataGred <= 4) return 8
  if (mataGred <= 7) return 5
  if (mataGred <= 10) return 3
  if (mataGred <= 12) return 1
  return 0
}

export function getPathwayFitScore(result: PathwayResult): number {
  if (!result.eligible) return 0
  if (result.pathway === 'matric' && result.merit !== undefined) {
    return BASE_SCORE + PRESTIGE_BONUS + matricAcademicBonus(result.merit)
  }
  if (result.pathway === 'stpm' && result.mataGred !== undefined) {
    return BASE_SCORE + PRESTIGE_BONUS + stpmAcademicBonus(result.mataGred, result.trackId)
  }
  return BASE_SCORE + PRESTIGE_BONUS
}

// --- Public API ---

export function checkAllPathways(
  grades: Record<string, string>,
  coqScore: number
): PathwayResult[] {
  const results: PathwayResult[] = []

  // Matriculation tracks
  for (const track of MATRIC_TRACKS) {
    results.push(checkMatricTrack(track.id, grades, coqScore))
  }

  // STPM bidang
  for (const bidang of STPM_BIDANGS) {
    results.push(checkStpmBidang(bidang.id, grades))
  }

  return results
}

File: halatuju_api/apps/courses/pathways.py
"""
Pre-university pathway eligibility engine for Matriculation and STPM.

Pure Python — no Django dependencies. Ported from halatuju-web/src/lib/pathways.ts.
Uses backend grade keys (math, addmath, chem, etc.) not frontend keys.

Input: student's SPM grades dict + CoQ score.
Output: eligibility + merit/mata_gred per pathway track.
"""

# ── Grade Point Scales ───────────────────────────────────────────────────

# Matriculation merit scale (from matrikulasi.moe.gov.my calculator)
MATRIC_GRADE_POINTS = {
    'A+': 25, 'A': 24, 'A-': 23, 'B+': 22, 'B': 21,
    'C+': 20, 'C': 19, 'D': 18, 'E': 17, 'G': 0,
}

# STPM (Form 6) mata gred scale — lower is better
STPM_MATA_GRED = {
    'A+': 1, 'A': 1, 'A-': 2, 'B+': 3, 'B': 4,
    'C+': 5, 'C': 6, 'D': 7, 'E': 8, 'G': 9,
}

# Grade ordering for comparison (lower index = better grade)
GRADE_ORDER = ['A+', 'A', 'A-', 'B+', 'B', 'C+', 'C', 'D', 'E', 'G']


def is_credit(grade):
    """Credit = C or better (mata gred <= 6)."""
    mg = STPM_MATA_GRED.get(grade)
    return mg is not None and mg <= 6


def meets_min(grade, min_grade):
    """Check if grade meets or exceeds the minimum threshold."""
    try:
        grade_idx = GRADE_ORDER.index(grade)
        min_idx = GRADE_ORDER.index(min_grade)
    except ValueError:
        return False
    return grade_idx <= min_idx  # lower index = better grade


# ── Matriculation ────────────────────────────────────────────────────────

MATRIC_TRACKS = [
    {'id': 'sains', 'name': 'Science', 'name_ms': 'Sains', 'name_ta': 'அறிவியல்'},
    {'id': 'kejuruteraan', 'name': 'Engineering', 'name_ms': 'Kejuruteraan', 'name_ta': 'பொறியியல்'},
    {'id': 'sains_komputer', 'name': 'Computer Science', 'name_ms': 'Sains Komputer', 'name_ta': 'கணினி அறிவியல்'},
    {'id': 'perakaunan', 'name': 'Accounting', 'name_ms': 'Perakaunan', 'name_ta': 'கணக்கியல்'},
]

# Subject requirements per track (using backend keys)
MATRIC_REQUIREMENTS = {
    'sains': [
        {'subject_id': 'math', 'min_grade': 'B'},
        {'subject_id': 'addmath', 'min_grade': 'C'},
        {'subject_id': 'chem', 'min_grade': 'C'},
        {'subject_id': 'phy', 'min_grade': 'C', 'alternatives': ['bio']},
    ],
    'kejuruteraan': [
        {'subject_id': 'math', 'min_grade': 'B'},
        {'subject_id': 'addmath', 'min_grade': 'C'},
        {'subject_id': 'phy', 'min_grade': 'C'},
        # 4th: any elective with min C — handled by slot filling
    ],
    'sains_komputer': [
        {'subject_id': 'math', 'min_grade': 'C'},
        {'subject_id': 'addmath', 'min_grade': 'C'},
        {'subject_id': 'comp_sci', 'min_grade': 'C'},
        # 4th: any elective with min C — handled by slot filling
    ],
    'perakaunan': [
        {'subject_id': 'math', 'min_grade': 'C'},
        # 3 electives with min C — handled by slot filling
    ],
}

_MATRIC_TRACK_MAP = {t['id']: t for t in MATRIC_TRACKS}


def find_best_elective(grades, exclude_ids, min_grade):
    """Find the best unused subject meeting min_grade.

    Returns {'id': str, 'grade': str} or None.
    """
    best = None
    for subj_id, grade in grades.items():
        if subj_id in exclude_ids:
            continue
        if not meets_min(grade, min_grade):
            continue
        pts = MATRIC_GRADE_POINTS.get(grade, 0)
        if best is None or pts > best[2]:
            best = (subj_id, grade, pts)
    if best is None:
        return None
    return {'id': best[0], 'grade': best[1]}


def check_matric_track(track_id, grades, coq_score):
    """Check eligibility for a specific Matric track.

    Args:
        track_id: One of 'sains', 'kejuruteraan', 'sains_komputer', 'perakaunan'.
        grades: Dict mapping backend subject keys to grade strings.
        coq_score: Co-curricular score (0-10).

    Returns:
        Dict with pathway result.
    """
    track = _MATRIC_TRACK_MAP[track_id]
    reqs = MATRIC_REQUIREMENTS[track_id]
    used_subjects = []  # list of {'id': str, 'grade': str}
    used_ids = set()

    # Check fixed requirements
    for req in reqs:
        candidates = [req['subject_id']] + req.get('alternatives', [])
        found = False
        for subj_id in candidates:
            grade = grades.get(subj_id)
            if grade and meets_min(grade, req['min_grade']):
                used_subjects.append({'id': subj_id, 'grade': grade})
                used_ids.add(subj_id)
                found = True
                break
        if not found:
            has_subject = any(grades.get(sid) for sid in candidates)
            return {
                'pathway': 'matric',
                'track_id': track_id,
                'track_name': track['name'],
                'track_name_ms': track['name_ms'],
                'track_name_ta': track['name_ta'],
                'eligible': False,
                'merit': None,
                'mata_gred': None,
                'max_mata_gred': None,
                'reason': 'pathways.gradeTooLow' if has_subject else 'pathways.subjectMissing',
            }

    # Fill remaining slots with best electives (always need 4 total)
    slots_needed = 4 - len(used_subjects)
    for _ in range(slots_needed):
        elective = find_best_elective(grades, used_ids, 'C')
        if elective is None:
            return {
                'pathway': 'matric',
                'track_id': track_id,
                'track_name': track['name'],
                'track_name_ms': track['name_ms'],
                'track_name_ta': track['name_ta'],
                'eligible': False,
                'merit': None,
                'mata_gred': None,
                'max_mata_gred': None,
                'reason': 'pathways.notEnoughElectives',
            }
        used_subjects.append(elective)
        used_ids.add(elective['id'])

    # Calculate merit
    subject_points = sum(
        MATRIC_GRADE_POINTS.get(s['grade'], 0) for s in used_subjects
    )
    academic = (subject_points / 100) * 90
    coq = min(max(coq_score, 0), 10)
    merit = min(academic + coq, 100)
    merit = round(merit * 100) / 100  # 2 decimal places

    return {
        'pathway': 'matric',
        'track_id': track_id,
        'track_name': track['name'],
        'track_name_ms': track['name_ms'],
        'track_name_ta': track['name_ta'],
        'eligible': True,
        'merit': merit,
        'mata_gred': None,
        'max_mata_gred': None,
        'reason': None,
    }


# ── STPM (Form 6) ───────────────────────────────────────────────────────

# Subject groups — student must have credits from 3 DIFFERENT groups
STPM_SCIENCE_GROUPS = [
    ['math', 'addmath'],
    ['phy'],
    ['chem'],
    ['bio'],
    ['eng_draw', 'eng_mech', 'eng_civil', 'eng_elec', 'reka_cipta',
     'sports_sci', 'srt', 'comp_sci', 'gkt'],
]

STPM_SOCSCI_GROUPS = [
    ['bm'],
    ['eng'],
    ['hist'],
    ['geo', 'psv'],
    ['islam', 'moral'],
    ['math', 'addmath'],
    ['poa'],
    ['sci', 'addsci'],
    ['ekonomi', 'business', 'keusahawanan', 'sports_sci', 'srt', 'comp_sci',
     'gkt', 'pertanian'],
]

STPM_BIDANGS = [
    {'id': 'sains', 'name': 'Science', 'name_ms': 'Sains',
     'name_ta': 'அறிவியல்', 'max_mata_gred': 18},
    {'id': 'sains_sosial', 'name': 'Social Science', 'name_ms': 'Sains Sosial',
     'name_ta': 'சமூக அறிவியல்', 'max_mata_gred': 18},
]

_STPM_BIDANG_MAP = {b['id']: b for b in STPM_BIDANGS}


def check_stpm_bidang(bidang_id, grades):
    """Check eligibility for a specific STPM bidang.

    Args:
        bidang_id: 'sains' or 'sains_sosial'.
        grades: Dict mapping backend subject keys to grade strings.

    Returns:
        Dict with pathway result.
    """
    bidang = _STPM_BIDANG_MAP[bidang_id]
    groups = STPM_SCIENCE_GROUPS if bidang_id == 'sains' else STPM_SOCSCI_GROUPS

    def _fail(reason, mata_gred=None, **extra):
        return {
            'pathway': 'stpm',
            'track_id': bidang_id,
            'track_name': bidang['name'],
            'track_name_ms': bidang['name_ms'],
            'track_name_ta': bidang['name_ta'],
            'eligible': False,
            'merit': None,
            'mata_gred': mata_gred,
            'max_mata_gred': bidang['max_mata_gred'],
            'reason': reason,
        }

    # General requirement: credit in BM
    bm_grade = grades.get('bm')
    if not bm_grade or not is_credit(bm_grade):
        return _fail('pathways.bmCreditRequired')

    # Find best credit from each group
    candidates = []  # (group_idx, subject_id, mata_gred)
    for gi, group in enumerate(groups):
        best_in_group = None
        for subj_id in group:
            grade = grades.get(subj_id)
            if not grade or not is_credit(grade):
                continue
            mg = STPM_MATA_GRED[grade]
            if best_in_group is None or mg < best_in_group[2]:
                best_in_group = (gi, subj_id, mg)
        if best_in_group:
            candidates.append(best_in_group)

    # Sort by mata gred (lowest first = best)
    candidates.sort(key=lambda c: c[2])

    if len(candidates) < 3:
        return _fail('pathways.notEnoughCredits')

    # Take best 3
    best_3 = candidates[:3]
    total_mata_gred = sum(c[2] for c in best_3)

    if total_mata_gred > bidang['max_mata_gred']:
        return _fail('pathways.mataGredTooHigh', mata_gred=total_mata_gred)

    return {
        'pathway': 'stpm',
        'track_id': bidang_id,
        'track_name': bidang['name'],
        'track_name_ms': bidang['name_ms'],
        'track_name_ta': bidang['name_ta'],
        'eligible': True,
        'merit': None,
        'mata_gred': total_mata_gred,
        'max_mata_gred': bidang['max_mata_gred'],
        'reason': None,
    }


# ── Public API ───────────────────────────────────────────────────────────

def check_all_pathways(grades, coq_score):
    """Check eligibility for all pre-university pathways.

    Args:
        grades: Dict mapping backend subject keys to grade strings.
        coq_score: Co-curricular score (0-10).

    Returns:
        List of 6 result dicts (4 matric + 2 stpm).
    """
    results = []

    for track in MATRIC_TRACKS:
        results.append(check_matric_track(track['id'], grades, coq_score))

    for bidang in STPM_BIDANGS:
        results.append(check_stpm_bidang(bidang['id'], grades))

    return results
